beps/bep_0056.rst

```diff
@@ -0,0 +1,145 @@
+(for seeding) and uncompressed files (for own usage).
+
+Most users prefer to remove such torrents, thus harming proper file
+distribution. For example: Organizations using Bittorrent for software
+distribution needs to have centralized storage for new customers, no
+matter how many customers have the same software already.
+
+Extension header
+================
+
+This extension uses the extension protocol (specified in `BEP 0010`_)
+to advertise client capability of using chunk compression. It defines
+following items in the extension protocol handshake message:
++-------+-----------------------------------------------------------+
+| name  | description                                               |
++=======+===========================================================+
+| c     | Dictionary of supported compression algorithms which maps |
+|       | its identifiers to its priority (unsigned 8-bit integer), |
+|       | clients can adjust it based on compression speed/ratio,   |
+|       | hardware support, performance, and power mode et cetera.  |
+|       | Priority set to zero means that the compression algorithm |
+|       | is not supported or disabled by user, the client must     |
+|       | ignore unknown algorithms.                                |
++-------+-----------------------------------------------------------+
+
+
+
+The compression algorithm is selected by taking the dictionary item with
+highest priority from intersection of items supported by both peers,
+if there isn't any suitable compression algorithm - compression will be disabled.
+
+Example of extension handshake message:
+
+::
+
+  {
+    'c': {
+      'p_zstd': 255,
+      's_zstd': 153,
+      'p_lz4': 106,
+      'p_density': 70,
+      's_lz4': 41,
+      's_density': 37
+    }
+  }
+
+
+Compression methods
+===================
+Extension provides two approaches (methods) to compression, which have
+their own trade-offs, so choice between these should be made by clients
+on per-torrent basis, using its metadata (properties like piece size).
+
+With **by-piece compression** method, client must compress each piece
+individually, which lowers overall compression ratio but result can
+be stored in cache and reused, probably providing more efficiency.
+If the client is caching compressed pieces in memory, then it can be
+decompressed when saving to disk or sending to peer, which not supports
+compression. To reduce piece re-compression, client should raise
+current algorithm's priority during handshake. This method has low
+efficiency with pieces smaller than 4 MB.
+
+Clients using **stream compression** method instead compresses whole
+data stream, so compression ratio should be higher. During handshake,
+clients should lower or raise algorithm's priority depending on expected
+factors that could impact compression efficiency and performance. This
+method can introduce performance issues if used on thousands of
+simultaneous connections.
```

Comment:
How do you synchronize *which byte* to start stream compression at?
Sending messages is asynchronous, in both directions. By the time I receive this handshake, I may have already sent other messages. In fact, I'm quite likely to.

I think you would need a message indicating that everything past it is compressed, and you probably ought to include which compression algorithm you picked in this message as well.

Reply:
Done with `cresponse` message